File: src/accessiweather/alert_lifecycle.py
"""Alert lifecycle intelligence: diff two alert snapshots to detect changes."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime, timedelta
from enum import Enum

from accessiweather.constants import SEVERITY_PRIORITY_MAP
from accessiweather.models.alerts import WeatherAlert, WeatherAlerts


class AlertChangeKind(Enum):
    NEW = "new"
    UPDATED = "updated"
    ESCALATED = "escalated"
    EXTENDED = "extended"
    CANCELLED = "cancelled"


@dataclass
class AlertChange:
    """Represents a single alert change (new, updated, escalated, extended, or cancelled)."""

    kind: AlertChangeKind
    alert: WeatherAlert | None = None
    alert_id: str = ""
    title: str = ""
    old_severity: str | None = None
    new_severity: str | None = None

    @property
    def is_severity_upgrade(self) -> bool:
        """Return True when the alert severity increased to a higher priority."""
        if self.old_severity is None or self.new_severity is None:
            return False
        old_priority: int = SEVERITY_PRIORITY_MAP.get(self.old_severity.lower(), 0)
        new_priority: int = SEVERITY_PRIORITY_MAP.get(self.new_severity.lower(), 0)
        return bool(new_priority > old_priority)


@dataclass
class AlertLifecycleDiff:
    """Structured diff between two alert snapshots."""

    new_alerts: list[AlertChange] = field(default_factory=list)
    updated_alerts: list[AlertChange] = field(default_factory=list)
    escalated_alerts: list[AlertChange] = field(default_factory=list)
    extended_alerts: list[AlertChange] = field(default_factory=list)
    cancelled_alerts: list[AlertChange] = field(default_factory=list)
    summary: str = "No changes"

    @property
    def has_changes(self) -> bool:
        """Return True if any alerts changed."""
        return bool(
            self.new_alerts
            or self.updated_alerts
            or self.escalated_alerts
            or self.extended_alerts
            or self.cancelled_alerts
        )


def compute_lifecycle_labels(alerts: list[WeatherAlert]) -> dict[str, str]:
    """
    Build alert_id → label mapping using messageType for NWS alerts.

    VisualCrossing alerts have no messageType equivalent — no label returned for them.

    Args:
        alerts: List of current active WeatherAlert objects.

    Returns:
        A dict mapping alert IDs to their display label strings.

    """
    labels: dict[str, str] = {}
    for alert in alerts:
        alert_id = alert.get_unique_id()
        if alert.source != "NWS" or not alert.message_type:
            continue  # VC and unknown sources: no label
        mt = alert.message_type.lower()
        if mt == "alert":
            labels[alert_id] = "New"
        elif mt == "update":
            labels[alert_id] = "Updated"
        # Cancel: alert is no longer active, omit
    return labels


def _alert_label(count: int, singular: str, plural: str | None = None) -> str:
    if plural is None:
        plural = singular + "s"
    return f"{count} {singular if count == 1 else plural}"


def _source_requires_cancel_confirmation(source: str | None) -> bool:
    """Return True for sources where disappearance alone is not trustworthy."""
    normalized_source = (source or "").strip().lower()
    return normalized_source in {"nws", "pirateweather"}


def _expires_extended(prev: WeatherAlert, curr: WeatherAlert) -> bool:
    """Return True when the expiry timestamp was pushed to a later time."""
    if prev.expires is None or curr.expires is None:
        return False
    prev_exp: datetime = prev.expires if prev.expires.tzinfo else prev.expires.replace(tzinfo=UTC)
    curr_exp: datetime = curr.expires if curr.expires.tzinfo else curr.expires.replace(tzinfo=UTC)
    return curr_exp > prev_exp


def _build_summary(
    new: list[AlertChange],
    updated: list[AlertChange],
    escalated: list[AlertChange],
    extended: list[AlertChange],
    cancelled: list[AlertChange],
) -> str:
    parts: list[str] = []

    if new:
        parts.append(_alert_label(len(new), "new alert"))

    if updated:
        # Check for legacy severity-upgrade entries in updated (kept for safety)
        upgrades = [c for c in updated if c.is_severity_upgrade]
        if upgrades:
            highest = max(
                upgrades,
                key=lambda c: SEVERITY_PRIORITY_MAP.get((c.new_severity or "").lower(), 0),
            )
            severity_note = (highest.new_severity or "higher").capitalize()
            label = _alert_label(len(updated), "updated")
            parts.append(f"{label} (severity upgraded to {severity_note})")
        else:
            parts.append(_alert_label(len(updated), "updated"))

    if escalated:
        highest = max(
            escalated,
            key=lambda c: SEVERITY_PRIORITY_MAP.get((c.new_severity or "").lower(), 0),
        )
        severity_note = (highest.new_severity or "higher").capitalize()
        label = _alert_label(len(escalated), "escalated")
        parts.append(f"{label} (to {severity_note})")

    if extended:
        parts.append(_alert_label(len(extended), "extended"))

    if cancelled:
        parts.append(_alert_label(len(cancelled), "cancelled"))

    return ", ".join(parts) if parts else "No changes"


def _alert_is_recently_issued(alert: WeatherAlert | None, max_age_minutes: int = 10) -> bool:
    """Return True only if the alert was issued within *max_age_minutes*."""
    if alert is None:
        return True  # no alert object → assume new to be safe
    effective_dt = alert.effective or alert.onset
    if effective_dt is None:
        return True  # no timestamp → assume new
    try:
        now = datetime.now(UTC)
        if effective_dt.tzinfo is None:
            effective_dt = effective_dt.replace(tzinfo=UTC)
        return (now - effective_dt) <= timedelta(minutes=max_age_minutes)
    except Exception:
        return True  # parse error → assume new to be safe


def diff_alerts(
    previous: WeatherAlerts | None,
    current: WeatherAlerts | None,
    confirmed_cancel_ids: set[str] | None = None,
) -> AlertLifecycleDiff:
    """
    Compare two alert snapshots and return a structured diff.

    Args:
        previous: The earlier snapshot (or None if no history).
        current:  The latest snapshot (or None if alerts are unavailable).
        confirmed_cancel_ids: Set of alert IDs confirmed cancelled via the NWS
            cancel endpoint. NWS alerts that disappear but are not in this set
            are silently suppressed (transient disappearance / Update supersession).
            If None, all NWS cancels are suppressed (safe default when the cancel
            endpoint could not be reached). Non-NWS alerts ignore this parameter.

    Returns:
        An AlertLifecycleDiff describing what changed.

    Change classification for alerts present in both snapshots:
    - ESCALATED: severity moved to a higher priority tier.
    - UPDATED: content hash, severity (downgrade), or urgency changed; not escalated.
    - EXTENDED: expiry timestamp pushed later with no other changes.
    - No-op: everything identical -- alert is omitted from all change lists.

    """
    prev_active = previous.get_active_alerts() if previous is not None else []
    curr_active = current.get_active_alerts() if current is not None else []

    prev_map: dict[str, WeatherAlert] = {a.get_unique_id(): a for a in prev_active}
    curr_map: dict[str, WeatherAlert] = {a.get_unique_id(): a for a in curr_active}

    new_changes: list[AlertChange] = []
    updated_changes: list[AlertChange] = []
    escalated_changes: list[AlertChange] = []
    extended_changes: list[AlertChange] = []
    cancelled_changes: list[AlertChange] = []

    # New: in current but not previous
    for alert_id, alert in curr_map.items():
        if alert_id not in prev_map:
            new_changes.append(
                AlertChange(
                    kind=AlertChangeKind.NEW,
                    alert=alert,
                    alert_id=alert_id,
                    title=alert.title or "",
                )
            )

    # Cancelled: in previous but not current
    for alert_id, prev_alert in prev_map.items():
        if alert_id not in curr_map:
            if _source_requires_cancel_confirmation(prev_alert.source):
                # NWS alerts: require explicit cancel confirmation via the NWS cancel endpoint.
                # Pirate Weather / WMO alerts: disappearance alone is not reliable enough to
                # announce a cancellation, and there is no authoritative cancel verification path.
                # If confirmed_cancel_ids is None (not fetched): suppress all such cancels.
                # If provided: only notify if this alert ID is explicitly confirmed cancelled.
                if confirmed_cancel_ids is not None and alert_id in confirmed_cancel_ids:
                    cancelled_changes.append(
                        AlertChange(
                            kind=AlertChangeKind.CANCELLED,
                            alert_id=alert_id,
                            title=prev_alert.title or "",
                        )
                    )
                # else: silently suppress (transient / Update supersession / stale cache)
            else:
                # Non-NWS sources (VisualCrossing, etc.): disappear = cancelled
                cancelled_changes.append(
                    AlertChange(
                        kind=AlertChangeKind.CANCELLED,
                        alert_id=alert_id,
                        title=prev_alert.title or "",
                    )
                )

    # Changed: in both maps -- classify into escalated / updated / extended
    for alert_id, alert in curr_map.items():
        if alert_id not in prev_map:
            continue  # already handled as NEW

        prev_alert = prev_map[alert_id]
        content_changed = alert.get_content_hash() != prev_alert.get_content_hash()
        severity_changed = alert.severity != prev_alert.severity
        urgency_changed = alert.urgency != prev_alert.urgency

        old_priority: int = SEVERITY_PRIORITY_MAP.get((prev_alert.severity or "").lower(), 0)
        new_priority: int = SEVERITY_PRIORITY_MAP.get((alert.severity or "").lower(), 0)
        severity_escalated = severity_changed and new_priority > old_priority

        if severity_escalated:
            escalated_changes.append(
                AlertChange(
                    kind=AlertChangeKind.ESCALATED,
                    alert=alert,
                    alert_id=alert_id,
                    title=alert.title or "",
                    old_severity=prev_alert.severity,
                    new_severity=alert.severity,
                )
            )
        elif content_changed or severity_changed or urgency_changed:
            updated_changes.append(
                AlertChange(
                    kind=AlertChangeKind.UPDATED,
                    alert=alert,
                    alert_id=alert_id,
                    title=alert.title or "",
                    old_severity=prev_alert.severity,
                    new_severity=alert.severity,
                )
            )
        elif _expires_extended(prev_alert, alert):
            extended_changes.append(
                AlertChange(
                    kind=AlertChangeKind.EXTENDED,
                    alert=alert,
                    alert_id=alert_id,
                    title=alert.title or "",
                )
            )

    # Bug fix 1: On first load (previous is None), suppress "New" notifications
    # for alerts that have been active for more than 10 minutes. They silently
    # enter the known-set without firing notifications.
    if previous is None:
        new_changes = [c for c in new_changes if _alert_is_recently_issued(c.alert)]

    summary = _build_summary(
        new_changes, updated_changes, escalated_changes, extended_changes, cancelled_changes
    )

    return AlertLifecycleDiff(
        new_alerts=new_changes,
        updated_alerts=updated_changes,
        escalated_alerts=escalated_changes,
        extended_alerts=extended_changes,
        cancelled_alerts=cancelled_changes,
        summary=summary,
    )

File: rust/crates/aw-core/src/alert_lifecycle.rs
//! Alert lifecycle: diff two alert snapshots and label NWS alerts.
//! Port of `accessiweather/alert_lifecycle.py`. The result types live in
//! [`crate::model`].

use std::collections::{BTreeMap, HashSet};

use chrono::{DateTime, Duration, Utc};

use crate::model::{AlertChange, AlertChangeKind, AlertLifecycleDiff, WeatherAlert, WeatherAlerts};

/// `SEVERITY_PRIORITY_MAP.get(severity.lower(), 0)`: unrecognised values rank
/// below "unknown" here (unlike `WeatherAlert::severity_priority`).
fn severity_rank(severity: Option<&str>) -> u8 {
    match severity.unwrap_or("").to_lowercase().as_str() {
        "unknown" => 1,
        "minor" => 2,
        "moderate" => 3,
        "severe" => 4,
        "extreme" => 5,
        _ => 0,
    }
}

impl AlertChange {
    fn new(kind: AlertChangeKind, alert_id: &str, alert: &WeatherAlert) -> Self {
        Self {
            kind,
            alert: None,
            alert_id: alert_id.to_string(),
            title: alert.title.clone(),
            old_severity: None,
            new_severity: None,
        }
    }

    /// True when the severity moved to a higher priority tier.
    pub fn is_severity_upgrade(&self) -> bool {
        match (&self.old_severity, &self.new_severity) {
            (Some(old), Some(new)) => severity_rank(Some(new)) > severity_rank(Some(old)),
            _ => false,
        }
    }
}

impl AlertLifecycleDiff {
    pub fn has_changes(&self) -> bool {
        !(self.new_alerts.is_empty()
            && self.updated_alerts.is_empty()
            && self.escalated_alerts.is_empty()
            && self.extended_alerts.is_empty()
            && self.cancelled_alerts.is_empty())
    }
}

/// `compute_lifecycle_labels`: alert id -> "New"/"Updated" from the NWS
/// `messageType`. Other sources get no label.
pub fn compute_lifecycle_labels(alerts: &[WeatherAlert]) -> BTreeMap<String, String> {
    let mut labels = BTreeMap::new();
    for alert in alerts {
        let Some(message_type) = alert.message_type.as_deref().filter(|m| !m.is_empty()) else {
            continue;
        };
        if alert.source.as_deref() != Some("NWS") {
            continue;
        }
        let label = match message_type.to_lowercase().as_str() {
            "alert" => "New",
            "update" => "Updated",
            _ => continue,
        };
        labels.insert(alert.unique_id(), label.to_string());
    }
    labels
}

fn count_label(count: usize, singular: &str) -> String {
    if count == 1 {
        format!("{count} {singular}")
    } else {
        format!("{count} {singular}s")
    }
}

/// Python's `str.capitalize()`: first character upper, the rest lower.
fn capitalize(s: &str) -> String {
    let mut chars = s.chars();
    match chars.next() {
        Some(first) => first
            .to_uppercase()
            .chain(chars.flat_map(char::to_lowercase))
            .collect(),
        None => String::new(),
    }
}

fn highest_new_severity(changes: &[AlertChange]) -> String {
    // Python's max() keeps the first of equal keys.
    let mut best = &changes[0];
    for change in &changes[1..] {
        if severity_rank(change.new_severity.as_deref())
            > severity_rank(best.new_severity.as_deref())
        {
            best = change;
        }
    }
    capitalize(
        best.new_severity
            .as_deref()
            .filter(|s| !s.is_empty())
            .unwrap_or("higher"),
    )
}

fn build_summary(diff: &AlertLifecycleDiff) -> String {
    let mut parts = Vec::new();
    if !diff.new_alerts.is_empty() {
        parts.push(count_label(diff.new_alerts.len(), "new alert"));
    }
    if !diff.updated_alerts.is_empty() {
        let label = count_label(diff.updated_alerts.len(), "updated");
        let upgrades: Vec<AlertChange> = diff
            .updated_alerts
            .iter()
            .filter(|c| c.is_severity_upgrade())
            .cloned()
            .collect();
        if upgrades.is_empty() {
            parts.push(label);
        } else {
            let note = highest_new_severity(&upgrades);
            parts.push(format!("{label} (severity upgraded to {note})"));
        }
    }
    if !diff.escalated_alerts.is_empty() {
        let note = highest_new_severity(&diff.escalated_alerts);
        let label = count_label(diff.escalated_alerts.len(), "escalated");
        parts.push(format!("{label} (to {note})"));
    }
    if !diff.extended_alerts.is_empty() {
        parts.push(count_label(diff.extended_alerts.len(), "extended"));
    }
    if !diff.cancelled_alerts.is_empty() {
        parts.push(count_label(diff.cancelled_alerts.len(), "cancelled"));
    }
    if parts.is_empty() {
        "No changes".into()
    } else {
        parts.join(", ")
    }
}

/// True when the alert was issued within `max_age_minutes` (or has no
/// timestamp, which is assumed new to be safe).
fn alert_is_recently_issued(
    alert: &WeatherAlert,
    now: DateTime<Utc>,
    max_age_minutes: i64,
) -> bool {
    match alert.effective.or(alert.onset) {
        None => true,
        Some(issued) => now - issued.with_timezone(&Utc) <= Duration::minutes(max_age_minutes),
    }
}

/// Active alerts keyed by unique id, keeping each id's first position and
/// last alert (Python dict-comprehension semantics).
fn active_by_id(alerts: Option<&WeatherAlerts>, now: DateTime<Utc>) -> Vec<(String, WeatherAlert)> {
    let mut out: Vec<(String, WeatherAlert)> = Vec::new();
    for alert in alerts.map(|a| a.active(now)).unwrap_or_default() {
        let id = alert.unique_id();
        match out.iter_mut().find(|(existing, _)| *existing == id) {
            Some(slot) => slot.1 = alert.clone(),
            None => out.push((id, alert.clone())),
        }
    }
    out
}

fn source_requires_cancel_confirmation(source: Option<&str>) -> bool {
    matches!(
        source.unwrap_or("").trim().to_lowercase().as_str(),
        "nws" | "pirateweather"
    )
}

/// `diff_alerts`: compare two snapshots.
///
/// `confirmed_cancel_ids` are the NWS cancel references; NWS / Pirate Weather
/// alerts that merely disappear are only reported cancelled when listed there
/// (`None` suppresses them all). On first load (`previous` is `None`) alerts
/// issued more than 10 minutes before `now` are not reported as new.
pub fn diff_alerts(
    previous: Option<&WeatherAlerts>,
    current: Option<&WeatherAlerts>,
    confirmed_cancel_ids: Option<&HashSet<String>>,
    now: DateTime<Utc>,
) -> AlertLifecycleDiff {
    let prev_map = active_by_id(previous, now);
    let curr_map = active_by_id(current, now);
    let find = |map: &[(String, WeatherAlert)], id: &str| -> Option<WeatherAlert> {
        map.iter().find(|(k, _)| k == id).map(|(_, a)| a.clone())
    };
    let mut diff = AlertLifecycleDiff::default();

    for (id, alert) in &curr_map {
        if find(&prev_map, id).is_none() {
            diff.new_alerts.push(AlertChange {
                alert: Some(alert.clone()),
                ..AlertChange::new(AlertChangeKind::New, id, alert)
            });
        }
    }

    for (id, prev_alert) in &prev_map {
        if find(&curr_map, id).is_some() {
            continue;
        }
        let announce = if source_requires_cancel_confirmation(prev_alert.source.as_deref()) {
            confirmed_cancel_ids.is_some_and(|ids| ids.contains(id))
        } else {
            true
        };
        if announce {
            diff.cancelled_alerts.push(AlertChange::new(
                AlertChangeKind::Cancelled,
                id,
                prev_alert,
            ));
        }
    }

    for (id, alert) in &curr_map {
        let Some(prev_alert) = find(&prev_map, id) else {
            continue;
        };
        let content_changed = alert.content_hash() != prev_alert.content_hash();
        let severity_changed = alert.severity != prev_alert.severity;
        let urgency_changed = alert.urgency != prev_alert.urgency;
        let escalated = severity_changed
            && severity_rank(Some(&alert.severity)) > severity_rank(Some(&prev_alert.severity));
        let with_severities = |kind| AlertChange {
            alert: Some(alert.clone()),
            old_severity: Some(prev_alert.severity.clone()),
            new_severity: Some(alert.severity.clone()),
            ..AlertChange::new(kind, id, alert)
        };
        if escalated {
            diff.escalated_alerts
                .push(with_severities(AlertChangeKind::Escalated));
        } else if content_changed || severity_changed || urgency_changed {
            diff.updated_alerts
                .push(with_severities(AlertChangeKind::Updated));
        } else if matches!((prev_alert.expires, alert.expires), (Some(p), Some(c)) if c > p) {
            diff.extended_alerts.push(AlertChange {
                alert: Some(alert.clone()),
                ..AlertChange::new(AlertChangeKind::Extended, id, alert)
            });
        }
    }

    if previous.is_none() {
        diff.new_alerts.retain(|c| {
            c.alert
                .as_ref()
                .is_none_or(|a| alert_is_recently_issued(a, now, 10))
        });
    }
    diff.summary = build_summary(&diff);
    diff
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::golden::{self, field};

    #[test]
    fn golden_diff_alerts() {
        let cases = golden::load("alerts/lifecycle.json");
        for case in cases.as_array().unwrap() {
            let now: DateTime<Utc> = field(case, "now");
            let previous: Option<WeatherAlerts> = field(case, "previous");
            let current: Option<WeatherAlerts> = field(case, "current");
            let ids: Option<HashSet<String>> = field(case, "confirmed_cancel_ids");
            let diff = diff_alerts(previous.as_ref(), current.as_ref(), ids.as_ref(), now);
            let expected: AlertLifecycleDiff = field(case, "diff");
            assert_eq!(diff, expected, "{}", case["name"]);
            assert_eq!(diff.has_changes(), case["has_changes"].as_bool().unwrap());
        }
    }

    #[test]
    fn golden_lifecycle_labels() {
        let case = golden::load("alerts/labels.json");
        let alerts: Vec<WeatherAlert> = field(&case, "alerts");
        let expected: BTreeMap<String, String> = field(&case, "labels");
        assert_eq!(compute_lifecycle_labels(&alerts), expected);
    }

    fn change(old: Option<&str>, new: Option<&str>) -> AlertChange {
        AlertChange {
            kind: AlertChangeKind::Updated,
            alert: None,
            alert_id: String::new(),
            title: String::new(),
            old_severity: old.map(Into::into),
            new_severity: new.map(Into::into),
        }
    }

    #[test]
    fn severity_upgrade_detection() {
        assert!(change(Some("Moderate"), Some("Extreme")).is_severity_upgrade());
        assert!(!change(Some("Extreme"), Some("Minor")).is_severity_upgrade());
        assert!(!change(Some("Severe"), Some("Severe")).is_severity_upgrade());
        assert!(!change(None, Some("Severe")).is_severity_upgrade());
        assert!(!change(Some("Severe"), None).is_severity_upgrade());
    }

    fn alert(id: &str, source: &str) -> WeatherAlert {
        WeatherAlert {
            id: Some(id.into()),
            source: Some(source.into()),
            ..WeatherAlert::new("T", "D")
        }
    }

    #[test]
    fn nws_cancel_requires_confirmation() {
        let now = Utc::now();
        let prev = WeatherAlerts {
            alerts: vec![alert("n1", "NWS")],
        };
        let empty = WeatherAlerts::default();
        let unconfirmed = diff_alerts(Some(&prev), Some(&empty), None, now);
        assert!(unconfirmed.cancelled_alerts.is_empty());
        let ids: HashSet<String> = ["n1".to_string()].into();
        let confirmed = diff_alerts(Some(&prev), Some(&empty), Some(&ids), now);
        assert_eq!(confirmed.cancelled_alerts.len(), 1);
        assert_eq!(confirmed.summary, "1 cancelled");
    }

    #[test]
    fn first_load_skips_old_alerts() {
        let now = Utc::now();
        let mut old = alert("old", "NWS");
        old.effective = Some((now - Duration::hours(2)).fixed_offset());
        let mut recent = alert("recent", "NWS");
        recent.effective = Some((now - Duration::minutes(5)).fixed_offset());
        let current = WeatherAlerts {
            alerts: vec![old, recent],
        };
        let diff = diff_alerts(None, Some(&current), None, now);
        assert_eq!(diff.new_alerts.len(), 1);
        assert_eq!(diff.new_alerts[0].alert_id, "recent");
    }
}
